drivers: dai: Add virtual DAI driver and tests

SurajSonawane2415 opened the pull request:
This PR is part of the [GSoC 2025 Project: Add Virtual DAI Component to SOF](https://wiki.linuxfoundation.org/gsoc/2025-gsoc-sound-open-firmware#project_2add_virtual_dai_component_to_sof) and is based on [SOF Issue #9349](https://github.com/thesofproject/sof/issues/9349).

We introduce a new `DAI_VIRTUAL` type with driver its associated support to enable audio pipelines that do **not rely on physical DAI hardware**. This is useful in the following scenarios:

* **Debugging and rapid prototyping**: Quickly create and test audio pipelines without needing actual hardware.
* **Software loopback preparation**: Lays the groundwork for a memory-to-memory audio pipeline within SOF.

## Key Changes
* Added **Virtual DAI driver** (initial version with basic logs).  
* Added **DTS binding** (`dai-virtual.yaml`).  
* Updated **Kconfig/CMake** and integrated with **i.MX8M DTS**.  
* Added **ztest-based test suite** for basic API validation.  

## Testing
* Platform: **i.MX8MP EVK** with Virtual DAI topology.  
* Environment: Zephyr-based SOF build.  
* Results:  
  - NoCodec card + Virtual DAI detected.  
  - Probe, config, and removal flows working as expected.
  - All test cases executed successfully, **all tests passed**

Logs:

```
*** Booting Zephyr OS build 6e3ea447cbfb ***
[00:00:00.046,014] <inf> main: SOF on imx8mp_evk
[00:00:00.051,043] <inf> main: SOF initialized
[00:00:00.055,909] <inf> ipc: ipc: new cmd 0x90030000
[00:00:00.061,749] <inf> ipc: ipc: new cmd 0x30100000
[00:00:00.066,973] <inf> pipe: pipeline new pipe_id 1 priority 0
[00:00:00.074,086] <inf> ipc: ipc: new cmd 0x30010000
[00:00:00.079,312] <inf> ipc: comp new 0x924549b4U type 2 id 1.4
[00:00:00.085,734] <inf> dai_comp: DAIZ: dai_new()
[00:00:00.090,941] <inf> dai_comp: DAIZ: dai_common_new()
[00:00:00.096,757] <inf> virtual_dai: virtual_dai_config_get
[00:00:00.102,833] <inf> virtual_dai: virtual_dai_probe
[00:00:00.108,476] <inf> dma: sof_dma_get: dir=0, cap=0x0, dev=0x0, flags=0x0
[00:00:00.116,028] <inf> dma: dma_init
[00:00:00.120,198] <inf> dma: sof_dma_get(): Final selection -> ID 0, sref=1, busy=0
[00:00:00.129,043] <inf> ipc: ipc: new cmd 0x80010000
[00:00:00.134,266] <inf> ipc: ipc: dai 16.7 -> config 
[00:00:00.139,820] <inf> virtual_dai: virtual_dai_config_get
[00:00:00.145,896] <inf> virtual_dai: virtual_dai_probe
[00:00:00.151,539] <inf> dai_comp: DAIZ: dai_set_config()
[00:00:00.157,354] <inf> virtual_dai: virtual_dai_config_set
[00:00:00.163,429] <inf> virtual_dai: virtual_dai_remove
[00:00:00.169,163] <inf> ipc: ipc_comp_dai_config() dai type = 16 index = 7
[00:00:00.176,541] <inf> ipc: comp:1.4 dai_config() dai type = 16 index = 7 dd 0x92c05ee0
[00:00:00.185,132] <inf> dai_comp: DAIZ: dai_get_handshake()
[00:00:00.191,208] <inf> virtual_dai: virtual_dai_get_properties
```

### Related Work

- **SOF PR**: [#10147 dai: Add support for Virtual DAI for rapid prototyping and software loopback](https://github.com/thesofproject/sof/pull/10147)  
  Adds Virtual DAI support on i.MX8MP with topology and `emul_dma` integration for software-based audio pipelines.

- **linux-imx PR**: [#30 ASoC: SOF: Add Virtual DAI support for i.MX8M](https://github.com/nxp-imx/linux-imx/pull/30)  
  Introduces `SOF_DAI_VIRTUAL` type and configuration, adds topology tokens, extends IPC3 parsing, and registers a Virtual DAI backend for i.MX8M. Enables prototyping and loopback pipelines without physical DAI hardware.

- **linux-imx nocodec feature commit**: Enabled NoCodec support for the i.MX8M platform to allow pipeline testing without real codecs.  
  Link: [nocodec-commit](https://github.com/SurajSonawane2415/linux-imx/commit/c82d84b2913f653eac9ec3712a7c79ca75d6e048)

### Future Work

* Implement playback and record:
   - Playback: log output frames and discard.
   - Record: generate frames filled with zeroes.
(Current limitation: To get `aplay` and `arecord` working, I replaced the real DMA with `emul_dma` for Virtual DAI. All functions execute, but data is not being copied in `dma_emul_work_handler` during playback.)

SurajSonawane2415 committed:
drivers: dai: Add virtual DAI driver

Introduce a virtual DAI driver for prototyping, loopback, and testing
without requiring physical DAI hardware.

This includes:
- Initial virtual DAI implementation (logging only)
- DTS binding (dai-virtual.yaml)
- Kconfig and CMake updates
- i.MX8M DTS integration

Signed-off-by: Suraj Sonawane <surajsonawane0215@gmail.com>

SurajSonawane2415 committed:
tests: drivers: dai: Add tests for virtual DAI

Add a basic ztest-based testsuite for the virtual DAI driver. This
includes initial API coverage and simple validation.

Signed-off-by: Suraj Sonawane <surajsonawane0215@gmail.com>

kv2019i commented on drivers/dai/virtual/virtual_dai.c:
```diff
@@ -0,0 +1,117 @@
+
+static int virtual_dai_probe(const struct device *dev)
+{
+	LOG_INF("%s", __func__);
```
I'd reduce the number of these LOG_INF()s in the merged version. I know this virtual DAI is more for R&D use, so logging can be on the more verbose side, but I'd still drop the function entry/exit logs (especially as Zephyr logging can add the function info with a build option, no need to add "__func__" in driver code.

kv2019i reviewed (changes requested):
Thanks @SurajSonawane24 ! Code is looking good. Only some minor notes on documentation, Kconfig defaults and log verbosity. Please see inline.

rruuaanng commented on dts/bindings/dai/dai-virtual.yaml:
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2025 Suraj Sonawane <surajsonawane0215@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Zephyr binding for a Virtual DAI (Digital Audio Interface).
+  This DAI is not connected to real hardware and is used for software-only
+  audio routing or testing within SOF + Zephyr systems.
```
```suggestion
title: Zephyr for a Virtual DAI (Digital Audio Interface)

description: |
  This DAI is not connected to real hardware and is used for software-only
  audio routing or testing within SOF + Zephyr systems.

```

SurajSonawane2415 replied:
Thank you for the suggestions, I’ve made the changes.

rruuaanng commented on drivers/dai/virtual/virtual_dai.c:
```diff
@@ -0,0 +1,117 @@
+	return 0;
+}
+
+static const struct dai_properties
+	*virtual_dai_get_properties(const struct device *dev, enum dai_dir dir, int stream_id)
```
```suggestion
static const struct dai_properties *virtual_dai_get_properties(const struct device *dev,
                                                                                         enum dai_dir dir, int stream_id)
```
or
```suggestion
static const struct dai_properties*
virtual_dai_get_properties(const struct device *dev, enum dai_dir dir, int stream_id)
```

SurajSonawane2415 replied:
@rruuaanng I accepted your suggestion, just kept the `*` with the function name (not the type) to follow kernel coding style.

SurajSonawane2415 commented:
> Thanks @SurajSonawane24 ! Code is looking good. Only some minor notes on documentation, Kconfig defaults and log verbosity. Please see inline.

@kv2019i Thanks for the detailed review!

- Dropped the redundant default n.
- Removed “SOF” mention from Kconfig description.
- Cleaned up logging by dropping function entry/exit logs, keeping only meaningful ones (using DBG instead of INF).

kv2019i reviewed (dismissed):
Thanks, looks good now @SurajSonawane2415 !

LaurentiuM1234 reviewed (changes requested):
Good start, but IMO this should still be in the draft stage. I would like to see some actual functionality being implemented here. IMO, as things stand now, this PR introduces a DAI driver that doesn't seem to do anything and a test suite that doesn't test anything particularly useful?

SurajSonawane2415 commented on dts/bindings/dai/dai-virtual.yaml:
```diff
@@ -0,0 +1,16 @@
+  Zephyr binding for a Virtual DAI (Digital Audio Interface).
+  This DAI is not connected to real hardware and is used for software-only
+  audio routing or testing within SOF + Zephyr systems.
+compatible: "virtual_dai"
```
Thank you I will make the changes.
